src/nvim/lua/executor.c

```diff
@@ -255,6 +257,117 @@ static struct luaL_Reg regex_meta[] = {
   { NULL, NULL }
 };
 
+// Dummy timer callback. Used by f_wait().
+static void dummy_timer_due_cb(TimeWatcher *tw, void *data)
+{
+}
+
+// Dummy timer close callback. Used by f_wait().
+static void dummy_timer_close_cb(TimeWatcher *tw, void *data)
+{
+  xfree(tw);
+}
+
+static bool nlua_wait_condition(lua_State *lstate, int *status,
+                                bool *callback_result)
+{
+  lua_pushvalue(lstate, 2);
+  *status = lua_pcall(lstate, 0, 1, 0);
+  if (*status) {
+    return true;  // break on error, but keep error on stack
+  }
+  *callback_result = lua_toboolean(lstate, -1);
+  lua_pop(lstate, 1);
+  return *callback_result;  // break if true
+}
+
+/// "vim.wait(timeout, condition[, interval])" function
+static int nlua_wait(lua_State *lstate)
+  FUNC_ATTR_NONNULL_ALL
+{
+  intptr_t timeout = luaL_checkinteger(lstate, 1);
+  if (timeout < 0) {
+    return luaL_error(lstate, "timeout must be > 0");
+  }
+
+  // Check if condition can be called.
+  //   Only error if we got an error on calling.
+  bool is_function = (lua_type(lstate, 2) == LUA_TFUNCTION);
+
+  // Run `condition` once, if possible.
+  int initial_status;
+  bool initial_callback_result;
+
+  nlua_wait_condition(lstate, &initial_status, &initial_callback_result);
+
+  // If calling errored
+  if (initial_status) {
+    // TODO(tjdevries): If I could find a way to use `vim.is_callable`
+    //        that'd be better. But I'm not sure how. Not a big deal atm.
+    if (!is_function) {
+      lua_pushliteral(lstate, "vim.wait: condition must be a function");
+      return lua_error(lstate);
+    }
+
+    lua_pushliteral(lstate, "vim.wait: `condition` callback raised error.");
+    return lua_error(lstate);
+  }
+
+  // Don't wait if the callback has already returned true.
+  if (initial_callback_result) {
+    lua_pushboolean(lstate, 1);
+    lua_pushnil(lstate);
+    return 2;
+  }
```

Comment:
We can check if condition is a callable table without trying calling condition.
```suggestion
  // Check if condition is callable table
  if (luaL_getmetafield(lstate, 2, "__call") != 0) {
    is_function = (lua_type(lstate, -1) == LUA_TFUNCTION);
    lua_pop(lstate, 1);
  }

  if (!is_function) {
    lua_pushliteral(lstate, "vim.wait: condition must be a function");
    return lua_error(lstate);
  }
```

Reply:
Changed this slightly, but did basically the same thing. Thanks, it's better this way.